File: services/taiwan-architect-kb-mcp/src/textrank.py
"""Lightweight Chinese TextRank keyword extraction, stdlib-only.

Approximates the approach used by TextRank4ZH (github.com/letiantian/TextRank4ZH):
build a co-occurrence graph over candidate terms within a sliding window, then
rank terms with iterative PageRank.

TextRank4ZH segments text into words with jieba before building the graph.
jieba (and the `textrank4zh` package itself) can't be installed in this
environment (no package registry access), so this module substitutes a
segmentation-free approximation: candidate terms are n-grams (default: 2-3
characters) over runs of CJK characters, filtered against a small stopword
set. This is strictly weaker than real word segmentation and should be
swapped for a jieba-backed implementation once dependency installation is
available.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_CJK_RUN_RE = re.compile(r"[一-鿿]+")

_STOPWORDS = {
    "的", "之", "與", "及", "或", "等", "於", "在", "為", "是", "者",
    "其", "而", "並", "以", "對", "由", "如", "則", "亦", "但", "又",
    "所", "本", "該", "此", "各", "應", "得", "均", "已", "未", "皆",
}


@dataclass(frozen=True)
class Keyword:
    term: str
    score: float


def _candidate_terms(text: str, ngram_sizes: tuple[int, ...] = (2, 3)) -> list[str]:
    terms: list[str] = []
    for run in _CJK_RUN_RE.findall(text):
        for size in ngram_sizes:
            for i in range(len(run) - size + 1):
                gram = run[i : i + size]
                if any(ch in _STOPWORDS for ch in gram):
                    continue
                terms.append(gram)
    return terms


def _build_graph(
    terms: list[str], window: int
) -> tuple[list[str], dict[str, dict[str, float]]]:
    graph: dict[str, dict[str, float]] = {}
    for i, term in enumerate(terms):
        graph.setdefault(term, {})
        for j in range(i + 1, min(i + window, len(terms))):
            other = terms[j]
            if other == term:
                continue
            graph.setdefault(other, {})
            graph[term][other] = graph[term].get(other, 0.0) + 1.0
            graph[other][term] = graph[other].get(term, 0.0) + 1.0
    return list(graph.keys()), graph


def _pagerank(
    nodes: list[str],
    graph: dict[str, dict[str, float]],
    damping: float = 0.85,
    max_iter: int = 100,
    tol: float = 1e-6,
) -> dict[str, float]:
    if not nodes:
        return {}
    n = len(nodes)
    scores = {node: 1.0 / n for node in nodes}
    out_weight = {
        node: sum(graph[node].values()) or 1.0 for node in nodes
    }

    for _ in range(max_iter):
        new_scores = {}
        max_delta = 0.0
        for node in nodes:
            rank_sum = 0.0
            for neighbor, weight in graph[node].items():
                rank_sum += (weight / out_weight[neighbor]) * scores[neighbor]
            new_scores[node] = (1 - damping) / n + damping * rank_sum
            max_delta = max(max_delta, abs(new_scores[node] - scores[node]))
        scores = new_scores
        if max_delta < tol:
            break
    return scores


def _merge_adjacent(keywords: list[str], original_terms: list[str]) -> list[str]:
    """Merge keyword n-grams that appear as consecutive overlapping spans in the
    source term sequence into longer phrases, mirroring TextRank4ZH's phrase
    extraction step."""
    keyword_set = set(keywords)
    merged: list[str] = []
    seen: set[str] = set()
    i = 0
    while i < len(original_terms):
        term = original_terms[i]
        if term in keyword_set and term not in seen:
            phrase = term
            j = i + 1
            while j < len(original_terms) and original_terms[j] in keyword_set:
                candidate = original_terms[j]
                if candidate.startswith(phrase[-(len(candidate) - 1) :]):
                    phrase += candidate[-1]
                    j += 1
                else:
                    break
            if phrase not in seen:
                merged.append(phrase)
                seen.add(phrase)
            i = j
        else:
            i += 1
    return merged


def extract_keywords(
    text: str, top_k: int = 10, window: int = 4
) -> list[Keyword]:
    """Extract up to `top_k` keywords/phrases from Chinese `text`."""
    terms = _candidate_terms(text)
    if not terms:
        return []

    nodes, graph = _build_graph(terms, window=window)
    scores = _pagerank(nodes, graph)

    ranked = sorted(scores.items(), key=lambda kv: kv[1], reverse=True)
    top_terms = [term for term, _ in ranked[: max(top_k * 3, top_k)]]
    phrases = _merge_adjacent(top_terms, terms)

    phrase_scores = []
    for phrase in phrases:
        # score a merged phrase as the max score of its constituent n-grams
        best = max(
            (scores[t] for t in top_terms if t in phrase or phrase in t),
            default=0.0,
        )
        phrase_scores.append(Keyword(term=phrase, score=best))

    phrase_scores.sort(key=lambda k: k.score, reverse=True)
    return phrase_scores[:top_k]

File: services/taiwan-architect-kb-mcp-cloudflare/src/lib/textrank.js
/**
 * Lightweight Chinese TextRank keyword extraction, zero dependencies.
 *
 * Direct JS port of the Python implementation in
 * services/taiwan-architect-kb-mcp/src/textrank.py — see that file's
 * module docstring for the full rationale (approximates TextRank4ZH
 * without jieba segmentation, using character n-grams instead).
 */

const CJK_RUN_RE = /[一-鿿]+/g;

const STOPWORDS = new Set([
  "的", "之", "與", "及", "或", "等", "於", "在", "為", "是", "者",
  "其", "而", "並", "以", "對", "由", "如", "則", "亦", "但", "又",
  "所", "本", "該", "此", "各", "應", "得", "均", "已", "未", "皆",
]);

function candidateTerms(text, ngramSizes = [2, 3]) {
  const terms = [];
  const runs = text.match(CJK_RUN_RE) || [];
  for (const run of runs) {
    for (const size of ngramSizes) {
      for (let i = 0; i <= run.length - size; i++) {
        const gram = run.slice(i, i + size);
        if ([...gram].some((ch) => STOPWORDS.has(ch))) continue;
        terms.push(gram);
      }
    }
  }
  return terms;
}

function buildGraph(terms, window) {
  const graph = new Map();
  const ensure = (term) => {
    if (!graph.has(term)) graph.set(term, new Map());
    return graph.get(term);
  };
  for (let i = 0; i < terms.length; i++) {
    const term = terms[i];
    ensure(term);
    for (let j = i + 1; j < Math.min(i + window, terms.length); j++) {
      const other = terms[j];
      if (other === term) continue;
      ensure(other);
      const a = ensure(term);
      const b = ensure(other);
      a.set(other, (a.get(other) || 0) + 1);
      b.set(term, (b.get(term) || 0) + 1);
    }
  }
  return graph;
}

function pagerank(graph, damping = 0.85, maxIter = 100, tol = 1e-6) {
  const nodes = [...graph.keys()];
  const n = nodes.length;
  if (n === 0) return new Map();

  let scores = new Map(nodes.map((node) => [node, 1 / n]));
  const outWeight = new Map(
    nodes.map((node) => {
      const total = [...graph.get(node).values()].reduce((a, b) => a + b, 0);
      return [node, total || 1];
    }),
  );

  for (let iter = 0; iter < maxIter; iter++) {
    const newScores = new Map();
    let maxDelta = 0;
    for (const node of nodes) {
      let rankSum = 0;
      for (const [neighbor, weight] of graph.get(node).entries()) {
        rankSum += (weight / outWeight.get(neighbor)) * scores.get(neighbor);
      }
      const value = (1 - damping) / n + damping * rankSum;
      newScores.set(node, value);
      maxDelta = Math.max(maxDelta, Math.abs(value - scores.get(node)));
    }
    scores = newScores;
    if (maxDelta < tol) break;
  }
  return scores;
}

function mergeAdjacent(keywords, originalTerms) {
  const keywordSet = new Set(keywords);
  const merged = [];
  const seen = new Set();
  let i = 0;
  while (i < originalTerms.length) {
    const term = originalTerms[i];
    if (keywordSet.has(term) && !seen.has(term)) {
      let phrase = term;
      let j = i + 1;
      while (j < originalTerms.length && keywordSet.has(originalTerms[j])) {
        const candidate = originalTerms[j];
        const suffix = phrase.slice(-(candidate.length - 1));
        if (candidate.startsWith(suffix)) {
          phrase += candidate[candidate.length - 1];
          j += 1;
        } else {
          break;
        }
      }
      if (!seen.has(phrase)) {
        merged.push(phrase);
        seen.add(phrase);
      }
      i = j;
    } else {
      i += 1;
    }
  }
  return merged;
}

/**
 * Extract up to `topK` keywords/phrases from Chinese `text`.
 * @returns {{term: string, score: number}[]}
 */
export function extractKeywords(text, topK = 10, window = 4) {
  const terms = candidateTerms(text);
  if (terms.length === 0) return [];

  const graph = buildGraph(terms, window);
  const scores = pagerank(graph);

  const ranked = [...scores.entries()].sort((a, b) => b[1] - a[1]);
  const topTerms = ranked.slice(0, Math.max(topK * 3, topK)).map(([term]) => term);
  const phrases = mergeAdjacent(topTerms, terms);

  const phraseScores = phrases.map((phrase) => {
    let best = 0;
    for (const t of topTerms) {
      if (phrase.includes(t) || t.includes(phrase)) {
        best = Math.max(best, scores.get(t) ?? 0);
      }
    }
    return { term: phrase, score: best };
  });

  phraseScores.sort((a, b) => b.score - a.score);
  return phraseScores.slice(0, topK);
}
